typescript/agentkit/src/action-providers/flaunch/flaunchActionProvider.ts

```diff
@@ -0,0 +1,638 @@
+   */
+  @CreateAction({
+    name: "flaunch",
+    description: `
+This tool allows launching a new memecoin using the flaunch protocol.
+
+It takes:
+- name: The name of the token
+- symbol: The symbol of the token
+- imageUrl: URL to the token image
+- description: Description of the token
+
+- websiteUrl: (optional) URL to the token website
+- discordUrl: (optional) URL to the token Discord
+- twitterUrl: (optional) URL to the token Twitter
+- telegramUrl: (optional) URL to the token Telegram
+
+Note:
+- If the optional fields are not provided, don't include them in the call.
+    `,
+    schema: FlaunchSchema,
+  })
+  async flaunch(
+    walletProvider: EvmWalletProvider,
+    args: z.infer<typeof FlaunchSchema>,
+  ): Promise<string> {
+    try {
+      const network = walletProvider.getNetwork();
+      const networkId = network.networkId;
+      const chainId = network.chainId;
+
+      if (!chainId || !networkId) {
+        throw new Error("Chain ID is not set.");
+      }
+
+      // upload image & token uri to ipfs
+      const tokenUri = await generateTokenUri(args.name, {
+        pinataConfig: { jwt: this.pinataJwt },
+        metadata: {
+          imageUrl: args.imageUrl,
+          description: args.description,
+          websiteUrl: args.websiteUrl,
+          discordUrl: args.discordUrl,
+          twitterUrl: args.twitterUrl,
+          telegramUrl: args.telegramUrl,
+        },
+      });
+
+      const data = encodeFunctionData({
+        abi: FAST_FLAUNCH_ZAP_ABI,
+        functionName: "flaunch",
+        args: [
+          {
+            name: args.name,
+            symbol: args.symbol,
+            tokenUri,
+            creator: walletProvider.getAddress(),
+          },
+        ],
+      });
+
+      const hash = await walletProvider.sendTransaction({
+        to: FastFlaunchZapAddress[chainId],
+        data,
+      });
+
+      const receipt = await walletProvider.waitForTransactionReceipt(hash);
+
+      const filteredPoolCreatedEvent = receipt.logs
+        .map(log => {
+          try {
+            if (
+              log.address.toLowerCase() !== FlaunchPositionManagerAddress[chainId].toLowerCase()
+            ) {
+              return null;
+            }
+
+            const event = decodeEventLog({
+              abi: POSITION_MANAGER_ABI,
+              data: log.data,
+              topics: log.topics,
+            });
+            return event.eventName === "PoolCreated" ? event.args : null;
+          } catch {
+            return null;
+          }
+        })
+        .filter((event): event is NonNullable<typeof event> => event !== null)[0];
+
+      const memecoinAddress = filteredPoolCreatedEvent._memecoin;
+      const chainSlug = Number(chainId) === base.id ? "base" : "base-sepolia";
+
+      return `Flaunched $${args.symbol} (${args.name}) with transaction hash: [${hash}](${NETWORK_ID_TO_VIEM_CHAIN[networkId].blockExplorers?.default.url}/tx/${hash})\n
```

Comment:
nit: From my experience, the AI responds a little better with tools that return JSON than return strings. I would suggest having this return a JSON.stringify of the results. The AI will transform this into an english response either way

Not necessary, just a recommendation